ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs

```diff
@@ -134,74 +120,72 @@ where
         ZeroKnowledgeProof::new(proof)
     }
 
-    pub fn gen_challenge_and_recurse<λ, I, J>(
+    pub fn gen_challenge_and_recurse<J, B>(
         proof_left: &GenericArray<F, TwoNMinusOne<λ>>,
         proof_right: &GenericArray<F, TwoNMinusOne<λ>>,
-        u: I,
-        v: J,
-    ) -> ProofGenerator<F>
+        uv_iterator: J,
+    ) -> Self
     where
-        λ: ArrayLength + Add + Sub<U1>,
+        λ: Add + Sub<U1>,
         <λ as Add>::Output: Sub<U1>,
         <<λ as Add>::Output as Sub<U1>>::Output: ArrayLength,
         <λ as Sub<U1>>::Output: ArrayLength,
-        I: IntoIterator<Item = F>,
-        J: IntoIterator<Item = F>,
-        I::IntoIter: ExactSizeIterator,
-        J::IntoIter: ExactSizeIterator,
+        J: Iterator<Item = B>,
+        B: Borrow<(GenericArray<F, λ>, GenericArray<F, λ>)>,
     {
-        let mut u = u.into_iter();
-        let mut v = v.into_iter();
-
-        debug_assert_eq!(u.len() % λ::USIZE, 0); // We should pad with zeroes eventually
-
-        let s = u.len() / λ::USIZE;
-
-        assert!(
-            s > 1,
-            "When the output is this small, you should validate the proof with a more straightforward reveal"
-        );
-
         let r: F = hash_to_field(
             &compute_hash(proof_left),
             &compute_hash(proof_right),
             λ::U128,
         );
-        let mut p = GenericArray::<F, λ>::generate(|_| F::ZERO);
-        let mut q = GenericArray::<F, λ>::generate(|_| F::ZERO);
+        let mut output = Vec::<(GenericArray<F, λ>, GenericArray<F, λ>)>::new();
         let denominator = CanonicalLagrangeDenominator::<F, λ>::new();
         let lagrange_table_r = LagrangeTable::<F, λ, U1>::new(&denominator, &r);
 
-        let pairs = (0..s).map(|_| {
-            for i in 0..λ::USIZE {
-                let x = u.next().unwrap_or(F::ZERO);
-                let y = v.next().unwrap_or(F::ZERO);
-                p[i] = x;
-                q[i] = y;
+        // iter and interpolate at x coordinate r
+        let mut index = 0;
+        let mut new_u_chunk = GenericArray::<F, λ>::generate(|_| F::ZERO);
+        let mut new_v_chunk = GenericArray::<F, λ>::generate(|_| F::ZERO);
+        for polynomial in uv_iterator {
+            let (u_chunk, v_chunk) = polynomial.borrow();
+            let u = lagrange_table_r.eval(u_chunk)[0];
+            let v = lagrange_table_r.eval(v_chunk)[0];
```

Comment:
I'd rather not call them "u" and "v", as those are the names of vectors and these are just points. I believe we are evaluating the polynomials p(x) and q(x) at the random challenge point "r" which is why I suggested p_r and q_r.